plugins/pur/core/src/main/java/org/pentaho/di/repository/pur/PurRepository.java

```diff
@@ -2121,7 +2314,12 @@ protected void saveDatabaseMeta( final RepositoryElementInterface element, final
       boolean isUpdate = element.getObjectId() != null;
       RepositoryFile file = null;
       if ( isUpdate ) {
-        file = pur.getFileById( element.getObjectId().getId() );
+        readWriteLock.readLock().lock();
+        try {
+          file = pur.getFileById( element.getObjectId().getId() );
+        } finally {
+          readWriteLock.readLock().unlock();
```

Comment:
The read lock should not be release between this and lower in the same method. The lock should be kept. Make this whole method wrapped in a single write lock.